src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs

```diff
@@ -47,16 +47,29 @@ public void BeginOperation(CancellationToken cancellationToken, Action<object?>
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            _awaitableState |= AwaitableState.Running;
-
             // Don't register if already completed, we would immediately unregistered in ObserveCancellation
             if (cancellationToken.CanBeCanceled && !IsCompleted)
             {
+                var previousAwaitableState = _awaitableState;
+
+                _cancellationTokenRegistration = cancellationToken.UnsafeRegister(callback, state);
+
+                // If we get back the default CancellationTokenRegistration then it means the
+                // callback synchronously and we can throw inline. This is safe because we haven't changed
+                // the state of the awaitable as yet.
+                if (_cancellationTokenRegistration == default(CancellationTokenRegistration))
+                {
+                    Debug.Assert(previousAwaitableState == _awaitableState, "The awaitable state changed!");
```

Comment:
Does the callback potentially change _awaitableState?  If yes, this isn't a valid assert, since this could happen.

Reply:
The assert only happens if _cancellationTokenRegistration is default. It's using that to know if it ran synchronously 😄

Reply:
Understood.  I'm wondering if, when it does run synchronously, if the callback might change the value of _awaitableState.

Reply:
It's not supposed to.